feat: add BlueskyPost

ascorbic opened the pull request:
This adds a Bluesky Post embed. It's styled like the official one (which is Preact and Tailwind), but is a fully static re-implementation.

delucis commented on demo/src/pages/bluesky.astro:
```diff
@@ -0,0 +1,26 @@
+		post="https://bsky.app/profile/wattenberger.com/post/3la54isd7az2i"
+	/>
+	<p>Video</p>
+	<BlueskyPost post="https://bsky.app/profile/mk.gg/post/3l7zx4hltw62a" />
```
Is there a bug rendering this one? There seems to be an extra empty container between the quote of Dan Abramov and the image below:

![image](https://github.com/user-attachments/assets/fa44acf5-1b1c-465b-af0c-eaef9fc313c8)

delucis commented on demo/src/pages/bluesky.astro:
```diff
@@ -0,0 +1,26 @@
+	<p>Video</p>
+	<BlueskyPost post="https://bsky.app/profile/mk.gg/post/3l7zx4hltw62a" />
+	<p>External media with quote</p>
+	<BlueskyPost post="https://bsky.app/profile/mk.gg/post/3la4wuiii4o2e" />
```
Ah this one too isn’t quite right I think:

![image](https://github.com/user-attachments/assets/aa176710-65ba-4a37-8719-98850046cb9a)

In the official embed the link renders inside the quoted post:

![image](https://github.com/user-attachments/assets/0c851e55-336b-4383-97e5-e8bfc7145eeb)

(The way they render the link here is weird anyway — just “github.com” — but at least the nesting helps communicate that the link is from the quoted post.)

delucis commented on docs/src/content/docs/components/bluesky.mdx:
Docs here look good. We should also document the styling API (which may involve deciding which if not all the custom properties are considered “public”). But that could also happen in a later PR if we want to get this tested in some real world scenarios before committing.

ascorbic replied:
I don't think they should be public. If people want to poke around and style them then they can, but I think we public API should just be to have them look like the official embed.

delucis reviewed (commented):
Thanks @ascorbic this looks great! 🙌 

I’m a bit nervous about containing the styles/avoiding a site’s styles from clashing, but don’t think the Declarative Shadow DOM approach we use in the `<BaselineStatus>` component is viable here, so we’ll have to roll with it. (This is partly why I’ve mostly chosen “low style” for the other embeds — even when Twitter API access was viable, we shipped very minimal styles. On the one hand that lets the content blend into a site’s styles, and on the other it got me off the hook from worrying about this kind of thing 😁)

Left a few questions and suggestions!

ascorbic commented on packages/astro-embed-bluesky/src/utils.ts:
```diff
@@ -0,0 +1,101 @@
+		atUri = `at://${did}/app.bsky.feed.post/${postId}`;
+	}
+
+	const hydratedPost = await agent.getPosts({ uris: [atUri] });
```
The AtpAgent allows you to pass in a fetch implementation, so that might be the best approach. It doesn't look like there's need to be a big change for that.

ascorbic commented:
Thanks for the review! I've refactored the styles quite a bit. There aren't any classes in styles.css now: it's just var declarations. Those are now all scoped to the post container, and are also prefixed. I've changed avatar to be a component. 

Those rendering issues were a nightmare! I thought I'd caught them all. They're all related to the way that Bluesky nests all of the records in different ways in different post types. I'll look at what's going on there.

ascorbic commented:
After a lot of refactoring it turns out the rendering issue wasn't related to the nesting logic at all: it was just that it was trying to next `a` tags, which the browser was helpfully "fixing". The refactor is an improvement though!

delucis commented on packages/astro-embed-bluesky/src/post.astro:
```diff
@@ -0,0 +1,157 @@
+
+		{embed && <Embed embed={embed} postUrl={postUrl} />}
+		<a href={postUrl} class="timestamp">
+			{formatter.format(new Date(record.createdAt ?? ''))}
```
Added some tests which failed in CI because of a different timezone compared to running locally which made me think again about this. I now remember we hit the same question with the old Twitter embed — if we render a timestamp on the server and therefore in the server’s time zone, that can be confusing because it is not local to the user (unlike the client-rendered components).

I think for Twitter I just decided to display a date with no time IIRC (still has a chance of being off, but it’s not as likely).

There’s also the option of some minimal JS hydration, e.g. to render something like this which updates the time to the user timezone:

```astro
<bluesky-time>
  <time datetime={record.createdAt?.toISOString()}>
    {formatter.format(new Date(record.createdAt ?? ''))}
  </time>
</bluesky-time>

<script>
const formatter = new Intl.DateTimeFormat('en-US', { dateStyle: 'long', timeStyle: 'short' });

customElements.define('bluesky-time', class BlueskyTime extends HTMLElement {
	connectedCallback() {
		try {
			const time = this.querySelector('time');
			const datetime = time?.getAttribute('datetime');
			if (!datetime) return;
			const date = new Date(datetime);
			time.textContent = formatter.format(date);
		} catch {}
	}
});
</script>
```

Does away with the zero-JS claim though.

ascorbic replied:
NEVER

delucis commented:
> The Netlify failure seems unrelated?

Yeah… probably my `safeGet()` refactor — a 404 like that should not error in theory

ascorbic commented:
Yeah, possibly just flakiness. If you try a redeploy it might work

delucis commented:
OK, curiously reverting [7699491](https://github.com/delucis/astro-embed/pull/167/commits/7699491ed3de73251a41d8320cce3084493c10e1) makes Netlify builds succeed again, but the build for that commit passed. I’d guess something about the re-use of the responses in the cached version perhaps triggered that `terminated` error? I guess we can do without the caching for now.

ascorbic commented:
Ah I think I know the problem. You can't re-use a `Response`, because as soon as you call `json()` or whatever, it consumes the stream. If you try to call it again you'll get a stream error. It would need to return a clone each time.

delucis commented:
> It would need to return a clone each time.

Yeah, that’s what my implementation was intending to be doing, but I probably messed it up somehow. I just tried an alternative approach using `ultrafetch`, which seems to work, so may make a follow-up PR to clean up some of that code.

ascorbic commented:
Yeah I think the problem was that unless I misunderstood it, it was only cloning the first time, and then caching the cloned copy and not cloning it again on each get.

delucis reviewed (approved):
OK, let’s do this!

I added one last change to let the post resolver fail gracefully instead of erroring out for unknown IDs. We can revisit fetch caching later.
